File: app/src/lib/player.ts
import { PIXEL_COUNT, type RGB, type Sequence } from '@/lib/types';
import { applyDisplay, renderCue } from '@/lib/effects';

export type PlayMode = 'auto' | 'manual';

/** Player state machine, per shared/DATA-MODEL.md section 4. */
export interface PlayerState {
  sequence: Sequence;
  mode: PlayMode;
  currentCue: number;
  cueStartMs: number;
  playing: boolean;
}

function blackout(): RGB[] {
  const frame: RGB[] = new Array(PIXEL_COUNT);
  for (let i = 0; i < PIXEL_COUNT; i++) frame[i] = [0, 0, 0];
  return frame;
}

export class Player {
  private sequence: Sequence;
  private mode: PlayMode = 'auto';
  private currentCue = 0;
  private cueStartMs = 0;
  private playing = false;
  // Loop flag (DATA-MODEL §4 "v2 loop flag"): when set, next() past the last cue wraps to 0
  // instead of stopping. Off by default; an explicit stop() still stops.
  private looping = false;
  // masterBrightness is a runtime global (BLE SET_BRIGHTNESS), default 255.
  private masterBrightness = 255;
  // Internal clock: the most recent time seen by tick(). Transitions anchor
  // cueStartMs to this, since their signatures (per the frozen contract) take
  // no `now` argument. A continuously-running tick loop keeps it fresh.
  private lastNow = 0;
  // Whether tick() has seen a real timestamp yet. Until it has, the frozen no-arg
  // transition methods can only anchor cueStartMs to a stale lastNow (0); the first
  // tick rebases so a command issued before the clock started can't skip cue 0.
  private clockStarted = false;

  constructor(sequence: Sequence) {
    this.sequence = sequence;
  }

  /** Readable snapshot of the current player state. */
  get state(): PlayerState {
    return {
      sequence: this.sequence,
      mode: this.mode,
      currentCue: this.currentCue,
      cueStartMs: this.cueStartMs,
      playing: this.playing,
    };
  }

  /** Runtime master brightness (0..255), applied in the display pipeline. */
  setBrightness(value: number): void {
    this.masterBrightness = value < 0 ? 0 : value > 255 ? 255 : value;
  }

  /** Loop flag: when set, the sequence wraps to cue 0 at the end instead of stopping. */
  setLoop(enabled: boolean): void {
    this.looping = enabled;
  }
  get loop(): boolean {
    return this.looping;
  }

  /** Set mode, reset to first cue, mark playing. */
  play(mode: PlayMode): void {
    this.mode = mode;
    this.currentCue = 0;
    this.cueStartMs = this.lastNow;
    this.playing = true;
  }

  /** Stop playback (output blackout). */
  stop(): void {
    this.playing = false;
  }

  /** Advance one cue; at the end wrap to 0 if looping, else stop. */
  next(): void {
    this.currentCue += 1;
    this.cueStartMs = this.lastNow;
    if (this.currentCue >= this.sequence.cues.length) {
      if (this.looping && this.sequence.cues.length > 0) {
        this.currentCue = 0; // loop flag: wrap instead of stopping
      } else {
        this.stop();
      }
    }
  }

  /** Step back one cue (min 0). */
  prev(): void {
    this.currentCue = Math.max(0, this.currentCue - 1);
    this.cueStartMs = this.lastNow;
  }

  /** Jump to a clamped cue index. */
  goto(i: number): void {
    const last = this.sequence.cues.length - 1;
    this.currentCue = last < 0 ? 0 : Math.min(Math.max(i, 0), last);
    this.cueStartMs = this.lastNow;
  }

  /** Advance the clock and return the display-pipeline frame (or blackout). */
  tick(nowMs: number): RGB[] {
    if (!this.clockStarted) {
      // First real timestamp: adopt it and rebase the current cue so elapsed starts
      // at 0 (matches the firmware, which stamps a real `now` on every transition).
      this.clockStarted = true;
      this.cueStartMs = nowMs;
    }
    this.lastNow = nowMs;
    if (!this.playing) return blackout();

    const cues = this.sequence.cues;
    if (cues.length === 0) {
      this.stop();
      return blackout();
    }

    if (this.mode === 'auto') {
      // Walk durations. next() re-anchors cueStartMs to nowMs, so elapsed resets
      // to 0 after each advance; the loop only continues through zero-duration
      // cues and terminates at end-of-sequence (which stops playback, unless looping).
      let advanced = 0;
      while (this.playing) {
        const cue = cues[this.currentCue];
        const elapsed = nowMs - this.cueStartMs;
        if (elapsed >= cue.durationMs) {
          this.next();
          // Guard: an all-zero-duration sequence under loop would spin forever. Cap advances at
          // cueCount per tick, then render whatever cue we landed on. (Never triggers for a
          // sequence with any positive-duration cue, so real content stays bit-identical.)
          if (++advanced > cues.length) break;
        } else {
          break;
        }
      }
      if (!this.playing) return blackout();
    }

    const cue = cues[this.currentCue];
    const elapsed = nowMs - this.cueStartMs;
    const frame = renderCue(cue, elapsed);
    return applyDisplay(frame, this.masterBrightness);
  }
}

File: firmware/lib/player/player.cpp
// player.cpp — transport / player state machine (AUTHORITATIVE CONTRACT: shared/DATA-MODEL.md §4).
// Portable C++ shared conceptually with the app's player.ts. Time is supplied by the caller;
// the player never reads a clock itself (keeps it host-testable).
//
// Display pipeline (per DATA-MODEL §3): out = gamma8( renderCue(cue, elapsed) * (master/255) ).
#include "player.h"
#include <cmath>

// round-half-up to [0,255] for the linear-domain master-brightness scale (before gamma).
// Matches the rounding convention used across the contract (see effects.cpp).
static inline uint8_t roundClamp(double v) {
  double r = std::floor(v + 0.5 + 1e-9);
  if (r < 0.0)   r = 0.0;
  if (r > 255.0) r = 255.0;
  return (uint8_t)r;
}

static inline void fillBlackout(RGB* frame) {
  for (uint16_t i = 0; i < PIXEL_COUNT; ++i) {
    frame[i].r = 0;
    frame[i].g = 0;
    frame[i].b = 0;
  }
}

Player::Player()
    : cues_(0),
      cueCount_(0),
      mode_(MODE_AUTO),
      currentCue_(0),
      cueStartMs_(0),
      playing_(false),
      masterBrightness_(255) {}

void Player::setSequence(const Cue* cues, uint16_t count) {
  cues_       = cues;
  cueCount_   = count;
  currentCue_ = 0;
  cueStartMs_ = 0;
  playing_    = false;   // reset transport to stopped at cue 0
}

// play → mode set, currentCue=0, cueStartMs=now, playing=true.
void Player::play(PlayMode mode, uint32_t nowMs) {
  mode_       = mode;
  currentCue_ = 0;
  cueStartMs_ = nowMs;
  playing_    = true;
}

// stop → playing=false (output blackout).
void Player::stop() {
  playing_ = false;
}

// next → currentCue+1, cueStartMs=now. If past last cue → stop() (v1: no loop).
void Player::next(uint32_t nowMs) {
  currentCue_ = (uint16_t)(currentCue_ + 1);
  cueStartMs_ = nowMs;
  if (currentCue_ >= cueCount_) {
    stop();
  }
}

// prev → currentCue-1 (min 0), cueStartMs=now.
void Player::prev(uint32_t nowMs) {
  if (currentCue_ > 0) {
    currentCue_ = (uint16_t)(currentCue_ - 1);
  }
  cueStartMs_ = nowMs;
}

// goto → currentCue=clamp(index, 0, cueCount-1), cueStartMs=now.
void Player::goto_(uint16_t index, uint32_t nowMs) {
  if (cueCount_ == 0) {
    currentCue_ = 0;
  } else if (index >= cueCount_) {
    currentCue_ = (uint16_t)(cueCount_ - 1);
  } else {
    currentCue_ = index;
  }
  cueStartMs_ = nowMs;
}

// tick → if !playing return blackout. elapsed = now - cueStartMs.
// In auto mode, if elapsed >= cue.durationMs → next() (may stop at end), then recompute.
// Manual mode never auto-advances. Returns the display-pipeline frame for the current cue.
void Player::tick(uint32_t nowMs, RGB* frame) {
  if (!playing_ || cues_ == 0 || cueCount_ == 0 || currentCue_ >= cueCount_) {
    fillBlackout(frame);
    return;
  }

  // Auto mode: walk cue durations. next() re-stamps cueStartMs_ = nowMs so elapsed
  // resets to 0 after each advance; the loop therefore only iterates past zero-duration
  // cues (skipped, never shown) and stops at end-of-sequence. Loop form (not a single
  // `if`) to stay bit-identical with app/src/lib/player.ts — see DATA-MODEL.md §4.
  if (mode_ == MODE_AUTO) {
    while (playing_ && (nowMs - cueStartMs_) >= cues_[currentCue_].durationMs) {
      next(nowMs);  // may stop() at end of sequence
      if (!playing_ || currentCue_ >= cueCount_) {
        fillBlackout(frame);
        return;
      }
    }
  }

  uint32_t elapsed = nowMs - cueStartMs_;
  RGB lin = renderCue(cues_[currentCue_], elapsed);

  // masterBrightness applied in the LINEAR domain, then gamma (DATA-MODEL §3).
  const double mScale = (double)masterBrightness_ / 255.0;
  RGB out;
  out.r = gamma8(roundClamp((double)lin.r * mScale));
  out.g = gamma8(roundClamp((double)lin.g * mScale));
  out.b = gamma8(roundClamp((double)lin.b * mScale));

  for (uint16_t i = 0; i < PIXEL_COUNT; ++i) {
    frame[i] = out;
  }
}

void Player::setMasterBrightness(uint8_t value) { masterBrightness_ = value; }
uint8_t Player::masterBrightness() const        { return masterBrightness_; }

bool     Player::isPlaying()  const { return playing_; }
PlayMode Player::mode()       const { return mode_; }
uint16_t Player::currentCue() const { return currentCue_; }
